src/lib/blockdev.c

```diff
@@ -267,17 +361,85 @@ gboolean bd_try_init (BDPluginSpec **require_plugins, BDUtilsLogFunc log_func, G
  */
 gboolean bd_reinit (BDPluginSpec **require_plugins, gboolean reload, BDUtilsLogFunc log_func, GError **error) {
     gboolean success = TRUE;
+    guint64 num_loaded = 0;
+
     g_mutex_lock (&init_lock);
     if (log_func && !bd_utils_init_logging (log_func, error))
         /* the error is already populated */
         success = FALSE;
 
-    if (!load_plugins (require_plugins, reload)) {
+    if (!load_plugins (require_plugins, reload, &num_loaded)) {
         g_set_error (error, BD_INIT_ERROR, BD_INIT_ERROR_PLUGINS_FAILED,
                      "Failed to load plugins");
-        /* the library is unusable without the plugins so we can just return here */
         success = FALSE;
-    }
+    } else
+        if (require_plugins && (*require_plugins == NULL) && reload)
+            /* requested to just unload all plugins */
+            success = (num_loaded == 0);
+
+    if (num_loaded == 0) {
+        if (require_plugins && (*require_plugins == NULL))
+            /* requested to load no plugins (NULL is the first item in the
+               array), none loaded -> OK */
+            initialized = TRUE;
+        else
+            initialized = FALSE;
+    } else
+        initialized = TRUE;
+
+    g_mutex_unlock (&init_lock);
+    return success;
+}
+
+/**
+ * bd_try_reinit:
+ * @require_plugins: (allow-none) (array zero-terminated=1): %NULL-terminated list
+ *                 of plugins that should be loaded (if no so_name is specified
+ *                 for the plugin, the default is used) or %NULL to load all
+ *                 plugins
+ * @reload: whether to reload the already loaded plugins or not
+ * @log_func: (allow-none) (scope notified): logging function to use or %NULL
+ *                                           to keep the old one
+ * @loaded_plugin_names: (allow-none) (out) (transfer container) (array zero-terminated=1): names of the successfully
+ *                                                                loaded plugins
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the library was successfully initialized or not
+ *
+ * If @reload is %TRUE all the plugins are closed and reloaded otherwise only
+ * the missing plugins are loaded.
+ *
+ * *UNLIKE IN CASE OF bd_init() AND bd_ensure_init(), FAILURE TO LOAD A PLUGIN
+ *  IS NOT CONSIDERED ERROR*
+ */
+gboolean bd_try_reinit (BDPluginSpec **require_plugins, gboolean reload, BDUtilsLogFunc log_func,
+                        gchar ***loaded_plugin_names, GError **error) {
+    gboolean success = TRUE;
+    guint64 num_loaded = 0;
+
+    g_mutex_lock (&init_lock);
+    if (log_func && !bd_utils_init_logging (log_func, error))
+        /* the error is already populated */
+        success = FALSE;
```

Comment:
This construction occurs in lots of places, but it always seems weird to me. Why continue if failure has already occurred? Why say that you have failed when you're going to keep on trying? Isn't the success of failure of initializing logging orthogonal to the success or failure of the actual task?